blog/content/post/2020-03-27-trenchboot-nlnet-lz.md

```diff
@@ -0,0 +1,419 @@
+    description = "Utility for reading, writing, erasing and verifying flash ROM chips";
+    license = licenses.gpl2;
+    platforms = platforms.all;
+  };
+}
+```
+
+5. Build `flashrom2`
+
+```bash
+nix-build -A flashrom2
+```
+
+> `nix-build` command builds any package which is added or already exists in
+`nixpkgs` library.
+
+6. Add `flashrom2` to NixOS profile to create symlinks.
+
+```bash
+nix-env -f . -iA flashrom2
+installing 'flashrom2-2.0'
+building '/nix/store/hbxi8l93x2qv3kzg7kjpfaa6pmkij48f-user-environment.drv'...
+created 21 symlinks in user environment
+```
+
+You have just built your first package in NixOS. Moreover, it is your custom
+build! However, as you can see entire process is non-trivial. Essential part
+each pakcage is `stdenv` library, which
+[documentation](https://nixos.org/nixos/nix-pills/fundamentals-of-stdenv.html)
+should clarify some issues. However, it is good to experiment with package
+manager on your own. Good point to start are already existing packages and their
+`default.nix` files.
+
+You may wonder why do we build our own tools, when most of them are available in
+default `nixpkgs`. Answer is simple - we have possibility to build not only
+tools, but also many other software and firmware components... You will
+learn more about it in next section.
+
+## Firmware
+
+As I mentioned in previous section, with NixOS we have possibility to customize
+our platform (both firmware and operating system). Before, we do that, I need to
+introduce some concepts. The better you understand them, the more awareness
+about security of your platform you will have and better understand the purpose
+of entire project. Let's break our entire system into following main parts. They
+are listed in order of execution when the platform is booted, so you can also
+see the flow and relations between them.
+
+#### BIOS
+
+Our solution is fully open-source, so we also use such firmware (BIOS) -
+**coreboot**. Also, as we are maintainers of coreboot for PC Engines platforms,
+we use
+[this](https://github.com/pcengines/coreboot/tree/pcengines_trenchboot_4.11.x)
+particular fork and branch. Every change appears there, so it is definitely
+place where you will find up-to-date firmware. I won't describe coreboot itself
+here. What you need to know, it is first code which is running during boot
+process. It initialize all hardware components on platform.
+
+#### GRUB
+
+When coreboot ends its work, a bootloader comes in. We use **GRUB**. Its task is
+to boot operating system. Normally, if NixOS is installed on hard disk, GRUB
+loads it to RAM. Being more precise, *Linux kernel* and *initrd* are being
+loaded. However, that operation is slightly different when we enable DRTM. As it
+is main goal of our project, let's take a look a little bit closer on it.
+
+We prepared extension module called `slaunch`. When DRTM is enabled, `slaunch`
+executes AMD's dedicated machine instruction `SKINIT`. Treat is 'pre-DRTM'
```

Comment:
Please double check if I wrote it correctly
```suggestion
executes AMD's dedicated machine instruction `SKINIT`. Treat is 'DLEvent'
```

Reply:
Actually, I wanted to define SKINIT as 'pre-DRTM'. It is simplification, but should be suitable for non-technical users too.